src/serial.rs

```diff
@@ -411,77 +448,15 @@ where
         config: impl Into<Config>,
         clocks: &Clocks,
     ) -> nb::Result<(), Infallible> {
-        // if we're currently busy transmitting, we have to wait until that is
-        // over -- regarding reception, we assume that the caller -- with
-        // exclusive access to the Serial instance due to &mut self -- knows
-        // what they're doing.
-        if self.usart.sr.read().tc().bit_is_clear() {
-            return nb::Result::Err(nb::Error::WouldBlock);
-        }
-        self.apply_config(config.into(), clocks);
-        nb::Result::Ok(())
-    }
-
-    /// Starts listening to the USART by enabling the _Received data
-    /// ready to be read (RXNE)_ interrupt and _Transmit data
-    /// register empty (TXE)_ interrupt
-    pub fn listen(&mut self, event: Event) {
-        self.usart.cr1.modify(|_, w| match event {
-            Event::Rxne => w.rxneie().set_bit(),
-            Event::Txe => w.txeie().set_bit(),
-            Event::Idle => w.idleie().set_bit(),
-        });
-    }
-
-    /// Stops listening to the USART by disabling the _Received data
-    /// ready to be read (RXNE)_ interrupt and _Transmit data
-    /// register empty (TXE)_ interrupt
-    pub fn unlisten(&mut self, event: Event) {
-        self.usart.cr1.modify(|_, w| match event {
-            Event::Rxne => w.rxneie().clear_bit(),
-            Event::Txe => w.txeie().clear_bit(),
-            Event::Idle => w.idleie().clear_bit(),
-        });
-    }
-
-    /// Returns true if the line idle status is set
-    pub fn is_idle(&self) -> bool {
-        self.usart.sr.read().idle().bit_is_set()
-    }
-
-    /// Returns true if the tx register is empty (and can accept data)
-    pub fn is_tx_empty(&self) -> bool {
-        self.usart.sr.read().txe().bit_is_set()
-    }
-
-    /// Returns true if the rx register is not empty (and can be read)
-    pub fn is_rx_not_empty(&self) -> bool {
-        self.usart.sr.read().rxne().bit_is_set()
-    }
-
-    /// Clear idle line interrupt flag
-    pub fn clear_idle_interrupt(&self) {
-        let _ = self.usart.sr.read();
-        let _ = self.usart.dr.read();
-    }
-
-    /// Returns ownership of the borrowed register handles
-    pub fn release(self) -> (USART, PINS) {
-        (self.usart, self.pins)
-    }
-
-    /// Separates the serial struct into separate channel objects for sending (Tx) and
-    /// receiving (Rx)
-    pub fn split(self) -> (Tx<USART, WORD>, Rx<USART, WORD>) {
-        (self.tx, self.rx)
+        reconfigure(&mut self.tx, &mut self.rx, config, clocks)
     }
 }
 
 macro_rules! hal {
     ($USARTX:ident) => {
         impl Instance for $USARTX {
-            fn ptr() -> *const uart_base::RegisterBlock {
-                <$USARTX>::ptr() as *const _
+            unsafe fn steal() -> Self {
+                unsafe { crate::pac::Peripherals::steal().$USARTX }
```

Comment:
Using of `steal` more then once is UB.

Reply:
Why using `steel` more than once is UB?
Isn't the use of constructions `*(<$USARTX>::ptr() as *const _)` and `crate::pac::Peripherals::steal().$USARTX` equivalent?

Reply:
Note: the solution to issue #406 was not implemented in these changes in order not to complicate the understanding of the changes made. This will need to be solved with a separate commit.

Reply:
We own a `USART` in two parts Tx and Rx (the issue with multithreaded access will be resolved later), so we do not store access to the `USART`, but get it as needed.

Reply:
The idea to use `steal` arose when it was necessary to return `USART` from the `release` function.

Reply:
I made the transition to using `steel` before I added `PinInfo`. Now it is possible to implement `USART` storage in `PinInfo` and rename `PinInfo` to `DataForRelease` or `ReleaseInfo`.